pkg/tfshim/sdk-v2/proposed_new.go

```diff
@@ -52,39 +47,189 @@ func htype2ctype(t hcty.Type) (cty.Type, error) {
+	switch {
+	case val.IsNull():
+		return cty.NullVal(ty)
+	case !val.IsKnown():
+		return cty.UnknownVal(ty)
+	case ty.Equals(cty.EmptyTuple):
+		return cty.EmptyTupleVal
+	case ty.Equals(cty.String):
+		return cty.StringVal(val.AsString())
+	case ty.Equals(cty.Number):
+		return cty.NumberVal(val.AsBigFloat())
+	case ty.Equals(cty.Bool):
+		if val.True() {
+			return cty.True
+		}
+		return cty.False
+	case ty.IsListType(), ty.IsSetType(), ty.IsTupleType():
+		l := val.LengthInt()
+		elems := make([]cty.Value, 0, l)
+		idx := 0
+		for it := val.ElementIterator(); it.Next(); {
+			_, ev := it.Element()
+			var eT cty.Type
+			if ty.IsTupleType() {
+				eT = ty.TupleElementType(idx)
+			} else {
+				eT = ty.ElementType()
+			}
+			newEv := hcty2ctyWithType(eT, ev)
+			elems = append(elems, newEv)
+			idx++
+		}
+		switch {
+		case ty.IsListType():
+			if len(elems) == 0 {
+				return cty.ListValEmpty(ty.ElementType())
+			}
+			return cty.ListVal(elems)
+		case ty.IsSetType():
+			if len(elems) == 0 {
+				return cty.SetValEmpty(ty.ElementType())
+			}
+			return cty.SetVal(elems)
+		case ty.IsTupleType():
+			return cty.TupleVal(elems)
+		default:
+			contract.Assertf(false, "match failure on cty.Type: unknown sequence type")
+		}
+	case ty.IsMapType():
+		eT := ty.ElementType()
+		elems := make(map[string]cty.Value)
+		for it := val.ElementIterator(); it.Next(); {
+			kv, ev := it.Element()
+			newEv := hcty2ctyWithType(eT, ev)
+			elems[kv.AsString()] = newEv
+		}
+		if len(elems) == 0 {
+			return cty.MapValEmpty(eT)
+		}
+		return cty.MapVal(elems)
+	case ty.IsObjectType():
+		atys := ty.AttributeTypes()
+		newAVs := make(map[string]cty.Value)
+		for name, aT := range atys {
+			av := val.GetAttr(name)
+			newAV := hcty2ctyWithType(aT, av)
+			newAVs[name] = newAV
+		}
+		return cty.ObjectVal(newAVs)
+	case ty.IsCapsuleType():
+		contract.Assertf(false, "Capsule types are not yet supported")
 	}
+	contract.Assertf(false, "match failure on hcty.Value: %v", val.GoString())
+	return cty.DynamicVal
+}
 
-	typeJSON, err := ctyjson.MarshalType(v.Type())
-	if err != nil {
-		return hcty.False, err
+func ctype2htype(t cty.Type) (hcty.Type, error) {
+	if t.GoString() == "cty.NilType" {
+		return hcty.NilType, nil
 	}
-	valueJSON, err := ctyjson.Marshal(v, v.Type())
+	typeJSON, err := ctyjson.MarshalType(t)
 	if err != nil {
-		return hcty.False, err
+		return hcty.Bool, err
 	}
-	typ, err := hctyjson.UnmarshalType(typeJSON)
-	if err != nil {
-		return hcty.False, err
+	return hctyjson.UnmarshalType(typeJSON)
+}
+
+func cty2hcty(val cty.Value) hcty.Value {
+	if val.IsKnown() && val.Equals(cty.NilVal).True() {
+		return hcty.NilVal
+	}
+	ty, err := ctype2htype(val.Type())
+	contract.AssertNoErrorf(err, "unexpected error in ctype2htype")
+	return cty2hctyWithType(ty, val)
+}
+
+func cty2hctyWithType(ty hcty.Type, val cty.Value) hcty.Value {
+	// Pattern-matching below is based on how hcty.Transform is written.
```

Comment:
What does "is based on" mean? Are we matching structure? the set of types? copy & paste?

Reply:
The most difficult thing, coming from OCaml languages, is making sure every branch is covered. I've looked at hcty.Transform as an example of covering every container branch. After that I've extended this to primitives (str/num/bool), and I think we do not need to support Capsule types in the Pulumi context. Those are all the cases I could come up with.